src/operator/controllers/intents_reconcilers/database/database_reconciler.go

```diff
@@ -0,0 +1,159 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ReasonApplyingDatabaseIntentsFailed     = "ApplyingDatabaseIntentsFailed"
+	ReasonAppliedDatabaseIntents            = "AppliedDatabaseIntents"
+	ReasonErrorFetchingPostgresServerConfig = "ErrorFetchingPostgresServerConfig"
+	ReasonMissingPostgresServerConfig       = "MissingPostgresServerConfig"
+)
+
+type DatabaseReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+	injectablerecorder.InjectableRecorder
+}
+
+func NewDatabaseReconciler(
+	client client.Client,
+	scheme *runtime.Scheme,
+) *DatabaseReconciler {
+	return &DatabaseReconciler{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clientIntents := &otterizev1alpha3.ClientIntents{}
+	logger := logrus.WithField("namespacedName", req.String())
+	err := r.client.Get(ctx, req.NamespacedName, clientIntents)
+	if err != nil && k8serrors.IsNotFound(err) {
+		logger.Info("No client intents found")
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, errors.Wrap(err)
+	}
+
+	if clientIntents.Spec == nil {
+		logger.Info("No specs found")
+		return ctrl.Result{}, nil
+	}
+
+	action := otterizev1alpha3.DBPermissionChangeApply
+	if !clientIntents.ObjectMeta.DeletionTimestamp.IsZero() {
+		action = otterizev1alpha3.DBPermissionChangeDelete
+	}
+
+	dbIntents := lo.Filter(clientIntents.GetCallsList(), func(intent otterizev1alpha3.Intent, _ int) bool {
+		return intent.Type == otterizev1alpha3.IntentTypeDatabase
+	})
+
+	dbInstanceToIntents := lo.GroupBy(dbIntents, func(intent otterizev1alpha3.Intent) string {
+		return intent.Name // "Name" is the db instance name in our case.
+	})
+	pgServerConfigs := otterizev1alpha3.PostgreSQLServerConfigList{}
+	err = r.client.List(ctx, &pgServerConfigs)
+	if err != nil {
+		r.RecordWarningEventf(clientIntents, ReasonErrorFetchingPostgresServerConfig,
+			"Error listing PostgresServerConfings. Error: %s", err.Error())
+		return ctrl.Result{}, errors.Wrap(err)
+	}
+
+	for databaseInstance, intents := range dbInstanceToIntents {
+		pgServerConf, err := findMatchingPGServerConfForDBInstance(databaseInstance, pgServerConfigs)
+		if err != nil {
+			r.RecordWarningEventf(clientIntents, ReasonMissingPostgresServerConfig,
+				"Could not find matching PostgreSQLServerConfig. Error: %s", err.Error())
+			return ctrl.Result{}, nil // Not returning error on purpose, missing PGServerConf - record event and move on
+		}
+		pgConfigurator := postgres.NewPostgresConfigurator(pgServerConf.Spec, r.client)
+		err = pgConfigurator.ConfigureDBFromIntents(ctx, clientIntents.GetServiceName(), clientIntents.Namespace, intents, action)
+		if err != nil {
+			r.RecordWarningEventf(clientIntents, ReasonApplyingDatabaseIntentsFailed,
+				"Failed applying database clientIntents: %s", err.Error())
+			return ctrl.Result{}, errors.Wrap(err)
+		}
+	}
```

Comment:
[Optional] Consider parallelizing this across DB instances (using errgroup). This comment applies in other places where you iterate over different DB instances.